fusion/FSL.md

```diff
@@ -0,0 +1,1220 @@
+instructions into alternative instruction forms. 
+
+Generally, this transformation process converts an N-tuple of processor
+instructions to an M-tuple of processor instructions. 
+
+The most commonly referenced transforms are fusion and fracture.
+
+When N > M holds, this is the case for fusion. 
+
+When N < M holds, this is the case for fracture. 
+
+Informally, fusion converts simpler operations into more complex operations
+with the intent of improving execution attributes. Fracture commonly replaces
+more complex operations with a set of simpler operations, with the 
+same intent of improving execution attributes.
+
+Often N-tuple instructions are elements of a standardized ISA and
+the M-tuple are customizations of the N-tuple set.  The M-tuple
+instructions are the result of transformation and are internal to the
+processor. 
+
+FSL assists the expression of the properties of these transformations and
+takes advantage of the bounds of the domain to provide an efficient syntax
+for these expressions.
+
+## High Level Operation
+
+Each N to M tuple mapping is expressed in a named FSL transform specification.
+
+The input to a transform is a sequence of instructions. The output of 
+a transform is a possibly transformed sequence. The input sequence can
+be modified in place or into a separate buffer. This is controlled by 
+FSL syntax.
+
+The operation of an FSL transform consists of a matching phase,
+a constraints checking phase, and the conversion phase.  The transform 
+phases are represented by three clauses in an FSL transform specification. 
+
+A phase returns an indication of success or failure. If a phase reports
+a failure, control is returned to the Fusion API. The API will continue
+processing and notify the system as required.
+
+The sequence clause declares the matching attributes to be applied to 
+the input sequence. The matching attributes are a list of
+instruction objects, either as assembly language or as a list of UIDs.
+
+The output of the sequence clause indicates which instructions
+are to be transformed, if any. If no match is found the sequence clause
+returns fail.
+
+In the simplest case, a positive match is indicated by an index and a length.
+The index is the position of the first matching instruction in the input
+sequence, length is the number of instructions to be transformed, inclusive of
+the index.
+
+A valid sequence indication is passed to the constraints clause. Subsequently,
+the constraints specification is applied to the indicated range and a 
+true/false flag is returned.
+
+If the constraints are satisfied, the constraints clause passes true to the
+conversion clause, along with the sequence indication.
+
+If the conversion clause receives true, it will apply the transformation
+on the input sequence beginning at the specified index and places the result 
+in an output sequence or modifies the input sequence in place as directed
+by syntax.
```

Comment:
Trying to understand the motivation behind allowing in-place modification. Can we have an example of where this will be preferred / useful?

Reply:
To minimize copies. e.g. take the case where there are no successful transforms performed, without in-place, an unmodified sequence of ops needs to be copied to the output.